pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/customobject/DistinctTable.java

```diff
@@ -0,0 +1,297 @@
+            dataTableBuilder.setColumn(i, (long) values[i]);
+            break;
+          case FLOAT:
+            dataTableBuilder.setColumn(i, (float) values[i]);
+            break;
+          case DOUBLE:
+            dataTableBuilder.setColumn(i, (double) values[i]);
+            break;
+          case STRING:
+            dataTableBuilder.setColumn(i, (String) values[i]);
+            break;
+          case BYTES:
+            dataTableBuilder.setColumn(i, (ByteArray) values[i]);
+            break;
+          // Add other distinct column type supports here
+          default:
+            throw new IllegalStateException();
+        }
+      }
+      dataTableBuilder.finishRow();
+    }
+    return dataTableBuilder.build().toBytes();
+  }
+
+  /*
+   * BROKER ONLY METHODS
+   */
+
+  /**
+   * Broker-side constructor to deserialize the {@code DistinctTable} from a {@code ByteBuffer}. The
+   * {@code DistinctTable} constructed this way cannot be used to add records or merge other {@code DistinctTable}s
+   * but can only be merged into the main {@code DistinctTable} because the order-by and limit information is missing.
+   */
+  public DistinctTable(ByteBuffer byteBuffer)
+      throws IOException {
+    DataTable dataTable = DataTableFactory.getDataTable(byteBuffer);
+    _dataSchema = dataTable.getDataSchema();
+    _limit = Integer.MIN_VALUE;
+    _uniqueRecords = null;
+    _sortedRecords = null;
+    int numRecords = dataTable.getNumberOfRows();
+    DataSchema.ColumnDataType[] columnDataTypes = _dataSchema.getColumnDataTypes();
+    int numColumns = columnDataTypes.length;
+    _records = new ArrayList<>(numRecords);
+    for (int i = 0; i < numRecords; i++) {
+      Object[] values = new Object[numColumns];
+      for (int j = 0; j < numColumns; j++) {
+        switch (columnDataTypes[j]) {
+          case INT:
+            values[j] = dataTable.getInt(i, j);
+            break;
+          case LONG:
+            values[j] = dataTable.getLong(i, j);
+            break;
+          case FLOAT:
+            values[j] = dataTable.getFloat(i, j);
+            break;
+          case DOUBLE:
+            values[j] = dataTable.getDouble(i, j);
+            break;
+          case STRING:
+            values[j] = dataTable.getString(i, j);
+            break;
+          case BYTES:
+            values[j] = dataTable.getBytes(i, j);
+            break;
+          // Add other distinct column type supports here
+          default:
+            throw new IllegalStateException();
+        }
+      }
+      _records.add(new Record(values));
+    }
+  }
+
+  /**
+   * (Broker-side) Merges another {@code DistinctTable} into the current one.
+   * <p>NOTE: {@code DistinctTable} on Broker-side has non-null {@code _records}.
+   */
+  public void brokerSideMerge(DistinctTable distinctTable) {
+    if (shouldNotAddMore()) {
+      return;
+    }
+    for (Record record : distinctTable._records) {
+      if (!add(record)) {
+        return;
```

Comment:
Same comment as above on serverSideMerge(). add() will not return false for order by queries so the check while merging each record is a penalty.